llm_studio/src/trl/trainer.py

```diff
@@ -0,0 +1,851 @@
+        Returns:
+            `tuple`: The input processed data.
+        """
+        for name, tensor_list in zip(
+            ["queries", "responses", "scores"], [queries, responses, scores]
+        ):
+            if not isinstance(tensor_list, list):
+                raise ValueError(
+                    f"{name} must be a list of tensors - got {type(tensor_list)}"
+                )
+            if not isinstance(tensor_list[0], torch.Tensor):
+                raise ValueError(
+                    f"Elements in {name} must be tensors - got {type(tensor_list[0])}"
+                )
+            if batch_size is not None and len(tensor_list) != batch_size:
+                raise ValueError(
+                    f"Batch size ({batch_size}) does not match number of examples"
+                    f" - but got {len(tensor_list)} for: {name}"
+                )
+
+        # add queries, scores and responses on the correct device
+        queries = [tensor.to(self.current_device) for tensor in queries]
+        responses = [tensor.to(self.current_device) for tensor in responses]
+        scores = [tensor.to(self.current_device) for tensor in scores]
+
+        # squeeze scores if needed
+        for i, score in enumerate(scores):
+            if score.dim() > 1:
+                raise ValueError(
+                    f"Scores must be 1-dimensional - got {score.dim()} for {score}"
+                )
+            elif score.dim() == 1:
+                scores[i] = score.squeeze()
+
+        return queries, responses, scores
+
+    def step(
+        self,
+        queries: List[torch.LongTensor],
+        responses: List[torch.LongTensor],
+        scores: List[torch.FloatTensor],
+    ):
+        """
+        Run a PPO optimisation step given a list of queries, model responses, and
+        rewards.
+
+        Args:
+            queries (List[`torch.LongTensor`]):
+                List of tensors containing the encoded queries of shape (`query_length`)
+            responses (List[`torch.LongTensor`]):
+                List of tensors containing the encoded responses of shape
+                (`response_length`)
+            scores (List[`torch.FloatTensor`]):
+                List of tensors containing the scores.
+
+        Returns:
+            `dict[str, Any]`: A summary of the training statistics
+        """
+        bs = self.cfg.training.batch_size
+
+        queries, responses, scores = self._step_safety_checker(
+            bs, queries, responses, scores
+        )
+
+        timing = dict()
+        t0 = time.time()
+
+        t = time.time()
+
+        model_inputs = self.prepare_model_inputs(queries, responses)
+
+        model_inputs_names = list(model_inputs.keys())
+
+        with torch.no_grad():
+            all_logprobs, _, values, masks = self.batched_forward_pass(
+                self.model, queries, responses, model_inputs
+            )
+            with self.model.backbone.disable_adapter():
```

Comment:
Awesome method to avoid having the reference model as a separate model.
We probably should fail if LORA is disabled or force enable lora with a warning in the logs.

Reply:
Also, force embedding gradients should be disabled.